opencog/cython/PythonEval.cc

```diff
@@ -1175,38 +1176,36 @@ void PythonEval::apply_as(const std::string& moduleFunction,
     PyGILState_Release(gstate);
 }
 
+// Check for errors in a script.
+void PythonEval::throw_on_error(){
+    if (PyErr_Occurred())
+    {
+        std::string errorString;
+        this->build_python_error_message(NO_FUNCTION_NAME, errorString);
+
+        // If there was an error, throw an exception so the user knows the
+        // script had a problem.
+        throw RuntimeException(TRACE_INFO, "%s", errorString.c_str());
+    }
+}
+
 std::string PythonEval::apply_script(const std::string& script)
 {
     std::lock_guard<std::recursive_mutex> lck(_mtx);
 
     // Grab the GIL
     PyGILState_STATE gstate = PyGILState_Ensure();
-
+    std::shared_ptr<void> _(nullptr, [&gstate](void * ptr){
+                                   // Release the GIL. No Python API allowed beyond this point.
```

Comment:
I don't recall, but maybe `std::unique_ptr` is faster/slimmer than `std::shared_ptr` ? Although it really pretty much doesn't matter here. 

Also, using a single underscore as a variable name is really, um, well, I get what you are doing, but this obfuscates the code. Call it `raii` because that is what it is, and future developers will read this and sai "Oh I know what RAII is" instead of puzzling over "what the heck is he trying to do here"?

I'm going to merge this anyway, but just thought I'd mention this because I'm nit-picky.

Reply:
You can also use `BOOST_SCOPE_EXIT`, see https://theboostcpplibraries.com/boost.scopeexit

Reply:
@linas Yea, i tried unique_ptr first, but it doesn't work. It turns out that according to standard shared_ptr should call destructor callback on null pointer and unique_ptr shouldn't. 
Updated with typedef for better clarity. Btw what is the issue with whitespaces??

Reply:
Ok, replaced shared_ptr with boost' scope guard. It seems it doesn't require adding new dependencies.